xds/src/main/java/io/grpc/xds/internal/sds/trust/SdsX509TrustManager.java

```diff
@@ -58,97 +59,76 @@
+            altNameFromCert, sanToVerifyMatcher.getExact(), sanToVerifyMatcher.getIgnoreCase());
+      case PREFIX:
+        return verifyDnsNamePrefix(
+            altNameFromCert, sanToVerifyMatcher.getPrefix(), sanToVerifyMatcher.getIgnoreCase());
+      case SUFFIX:
+        return verifyDnsNameSuffix(
+            altNameFromCert, sanToVerifyMatcher.getSuffix(), sanToVerifyMatcher.getIgnoreCase());
+      case CONTAINS:
+        return verifyDnsNameContains(
+            altNameFromCert, sanToVerifyMatcher.getContains(), sanToVerifyMatcher.getIgnoreCase());
+      case SAFE_REGEX:
+        return verifyDnsNameSafeRegex(altNameFromCert, sanToVerifyMatcher.getSafeRegex());
+      default:
+        throw new IllegalArgumentException(
+            "Unknown match-pattern-case " + sanToVerifyMatcher.getMatchPatternCase());
     }
-    // Wildcard pattern
+  }
 
-    // WILDCARD PATTERN RULES:
-    // 1. Asterisk (*) is only permitted in the left-most domain name label and must be the
-    //    only character in that label (i.e., must match the whole left-most label).
-    //    For example, *.example.com is permitted, while *a.example.com, a*.example.com,
-    //    a*b.example.com, a.*.example.com are not permitted.
-    // 2. Asterisk (*) cannot match across domain name labels.
-    //    For example, *.example.com matches test.example.com but does not match
-    //    sub.test.example.com.
-    // 3. Wildcard patterns for single-label domain names are not permitted.
+  private static boolean verifyDnsNameSafeRegex(
+          String altNameFromCert, RegexMatcher sanToVerifySafeRegex) {
+    Pattern safeRegExMatch = Pattern.compile(sanToVerifySafeRegex.getRegex());
+    return safeRegExMatch.matches(altNameFromCert);
+  }
 
-    if (!pattern.startsWith("*.") || pattern.indexOf('*', 1) != -1) {
-      // Asterisk (*) is only permitted in the left-most domain name label and must be the only
-      // character in that label
+  private static boolean verifyDnsNamePrefix(
+      String altNameFromCert, String sanToVerifyPrefix, boolean ignoreCase) {
+    if (Strings.isNullOrEmpty(sanToVerifyPrefix)) {
       return false;
     }
+    return ignoreCase
+        ? altNameFromCert.toLowerCase().startsWith(sanToVerifyPrefix.toLowerCase())
+        : altNameFromCert.startsWith(sanToVerifyPrefix);
+  }
 
-    // Optimization: check whether sanToVerify is too short to match the pattern. sanToVerify must
-    // be at
-    // least as long as the pattern because asterisk must match the whole left-most label and
-    // sanToVerify starts with a non-empty label. Thus, asterisk has to match one or more
-    // characters.
-    if (sanToVerify.length() < pattern.length()) {
-      // sanToVerify too short to match the pattern.
+  private static boolean verifyDnsNameSuffix(
+          String altNameFromCert, String sanToVerifySuffix, boolean ignoreCase) {
+    if (Strings.isNullOrEmpty(sanToVerifySuffix)) {
       return false;
```

Comment:
Why return false for empty suffix?

Reply:
See https://github.com/envoyproxy/envoy/blob/master/api/envoy/type/matcher/v3/string.proto#L47 (it should  be suffix and *not* prefix - copy-paste bug) and https://github.com/envoyproxy/envoy/blob/master/api/envoy/type/matcher/v3/string.proto#L52 which says `{min_len: 1}`.

Reply:
I see, thanks.